File: app/src/main/java/code/name/monkey/retromusic/ui/activities/SettingsActivity.java
package code.name.monkey.retromusic.ui.activities;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.afollestad.materialdialogs.color.ColorChooserDialog;
import com.google.android.material.appbar.AppBarLayout;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.transition.TransitionManager;
import butterknife.BindView;
import butterknife.ButterKnife;
import code.name.monkey.appthemehelper.ThemeStore;
import code.name.monkey.appthemehelper.util.ColorUtil;
import code.name.monkey.appthemehelper.util.ToolbarContentTintHelper;
import code.name.monkey.retromusic.R;
import code.name.monkey.retromusic.ui.activities.base.AbsBaseActivity;
import code.name.monkey.retromusic.ui.fragments.settings.MainSettingsFragment;
import code.name.monkey.retromusic.util.PreferenceUtil;

public class SettingsActivity extends AbsBaseActivity implements ColorChooserDialog.ColorCallback, SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = "SettingsActivity";

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.app_bar)
    AppBarLayout appBarLayout;

    @BindView(R.id.title)
    TextView title;

    @BindView(R.id.detail_content_frame)
    @Nullable
    FrameLayout detailsFrame;

    private FragmentManager fragmentManager = getSupportFragmentManager();

    @Override
    public void onColorSelection(@NonNull ColorChooserDialog dialog, @ColorInt int selectedColor) {
        switch (dialog.getTitle()) {
            case R.string.primary_color:
                int theme = ColorUtil.isColorLight(selectedColor) ?
                        PreferenceUtil.getThemeResFromPrefValue("light") :
                        PreferenceUtil.getThemeResFromPrefValue("dark");

                ThemeStore.editTheme(this).activityTheme(theme).primaryColor(selectedColor).commit();
                break;
            case R.string.accent_color:
                ThemeStore.editTheme(this).accentColor(selectedColor).commit();
                break;
        }

        recreate();
    }

    @Override
    public void onColorChooserDismissed(@NonNull ColorChooserDialog dialog) {

    }

    @Override
    protected void onCreate(@Nullable Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.activity_settings);
        ButterKnife.bind(this);

        setStatusbarColorAuto();
        setNavigationbarColorAuto();

        setLightNavigationBar(true);

        setupToolbar();

        if (bundle == null) {
            fragmentManager.beginTransaction().replace(R.id.content_frame, new MainSettingsFragment())
                    .commit();
        }
    }

    private void setupToolbar() {
        toolbar.setBackgroundColor(ThemeStore.primaryColor(this));
        appBarLayout.setBackgroundColor(ThemeStore.primaryColor(this));
        setSupportActionBar(toolbar);
        setTitle(null);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());
        title.setTextColor(ThemeStore.textColorPrimary(this));
        ToolbarContentTintHelper.colorBackButton(toolbar, ThemeStore.accentColor(this));
    }

    public void setupFragment(Fragment fragment, @StringRes int titleName) {
        FragmentTransaction fragmentTransaction = fragmentManager
                .beginTransaction()
                .setCustomAnimations(R.anim.sliding_in_left, R.anim.sliding_out_right, android.R.anim.slide_in_left, android.R.anim.slide_out_right);

        title.setText(titleName);

        if (detailsFrame == null) {
            fragmentTransaction.replace(R.id.content_frame, fragment, fragment.getTag());
            fragmentTransaction.addToBackStack(null);
            fragmentTransaction.commit();
        } else {
            fragmentTransaction.replace(R.id.detail_content_frame, fragment, fragment.getTag());
            fragmentTransaction.commit();
        }
    }

    @Override
    public void onBackPressed() {
        if (fragmentManager.getBackStackEntryCount() == 0) {
            super.onBackPressed();
        } else {
            title.setText(R.string.action_settings);
            fragmentManager.popBackStack();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void addAppbarLayoutElevation(float v) {
        TransitionManager.beginDelayedTransition(appBarLayout);
        appBarLayout.setElevation(v);
    }

    @Override
    public void onPause() {
        super.onPause();
        PreferenceUtil.getInstance().unregisterOnSharedPreferenceChangedListener(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        PreferenceUtil.getInstance().registerOnSharedPreferenceChangedListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Log.i(TAG, "onSharedPreferenceChanged: ");
        if (key.equals(PreferenceUtil.PROFILE_IMAGE_PATH)) {
            recreate();
        }
    }
}

File: app/src/main/java/code/name/monkey/retromusic/ui/activities/SettingsActivity.kt
package code.name.monkey.retromusic.ui.activities

import android.content.SharedPreferences
import android.os.Bundle
import android.view.MenuItem
import androidx.annotation.ColorInt
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import butterknife.ButterKnife
import code.name.monkey.appthemehelper.ThemeStore
import code.name.monkey.appthemehelper.util.ColorUtil
import code.name.monkey.appthemehelper.util.ToolbarContentTintHelper
import code.name.monkey.retromusic.R
import code.name.monkey.retromusic.ui.activities.base.AbsBaseActivity
import code.name.monkey.retromusic.ui.fragments.settings.MainSettingsFragment
import code.name.monkey.retromusic.util.PreferenceUtil
import com.afollestad.materialdialogs.color.ColorChooserDialog
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsActivity : AbsBaseActivity(), ColorChooserDialog.ColorCallback, SharedPreferences.OnSharedPreferenceChangeListener {

    private val fragmentManager = supportFragmentManager

    override fun onColorSelection(dialog: ColorChooserDialog, @ColorInt selectedColor: Int) {
        when (dialog.title) {
            R.string.primary_color -> {
                val theme = if (ColorUtil.isColorLight(selectedColor))
                    PreferenceUtil.getThemeResFromPrefValue("light")
                else
                    PreferenceUtil.getThemeResFromPrefValue("dark")

                ThemeStore.editTheme(this).activityTheme(theme).primaryColor(selectedColor).commit()
            }
            R.string.accent_color -> ThemeStore.editTheme(this).accentColor(selectedColor).commit()
        }
        recreate()
    }

    override fun onColorChooserDismissed(dialog: ColorChooserDialog) {

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        ButterKnife.bind(this)

        setStatusbarColorAuto()
        setNavigationbarColorAuto()

        setLightNavigationBar(true)

        setupToolbar()

        if (savedInstanceState == null) {
            fragmentManager.beginTransaction().replace(R.id.contentFrame, MainSettingsFragment())
                    .commit()
        }
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        title = null
        toolbar.setBackgroundColor(ThemeStore.primaryColor(this))
        appBarLayout.setBackgroundColor(ThemeStore.primaryColor(this))
        toolbar.setNavigationOnClickListener { onBackPressed() }
        settingsTitle.setTextColor(ThemeStore.textColorPrimary(this))
        ToolbarContentTintHelper.colorBackButton(toolbar, ThemeStore.accentColor(this))
    }

    fun setupFragment(fragment: Fragment, @StringRes titleName: Int) {
        val fragmentTransaction = fragmentManager
                .beginTransaction()
                .setCustomAnimations(R.anim.sliding_in_left, R.anim.sliding_out_right, android.R.anim.slide_in_left, android.R.anim.slide_out_right)

        settingsTitle.setText(titleName)

        if (detailContentFrame == null) {
            fragmentTransaction.replace(R.id.contentFrame, fragment, fragment.tag)
            fragmentTransaction.addToBackStack(null)
            fragmentTransaction.commit()
        } else {
            fragmentTransaction.replace(R.id.detailContentFrame, fragment, fragment.tag)
            fragmentTransaction.commit()
        }
    }

    override fun onBackPressed() {
        if (fragmentManager.backStackEntryCount == 0) {
            super.onBackPressed()
        } else {
            settingsTitle.setText(R.string.action_settings)
            fragmentManager.popBackStack()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    public override fun onPause() {
        super.onPause()
        PreferenceUtil.getInstance().unregisterOnSharedPreferenceChangedListener(this)
    }

    public override fun onResume() {
        super.onResume()
        PreferenceUtil.getInstance().registerOnSharedPreferenceChangedListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (key == PreferenceUtil.PROFILE_IMAGE_PATH) {
            recreate()
        }
    }

    companion object {
        const val TAG: String = "SettingsActivity"
    }
}
